driver-core/src/main/java/com/datastax/driver/core/CodecFactory.java

```diff
@@ -0,0 +1,250 @@
+                        ? null
+                        : cqlType.getTypeArguments().get(0);
+                    return (TypeCodec<T>)newSetCodec(codecRegistry.codecFor(elementType, set.iterator().next()));
+                }
+            }
+
+            if ((cqlType == null || cqlType.getName() == MAP) && value instanceof Map) {
+                Map map = (Map)value;
+                if (map.isEmpty()) {
+                    DataType keyType = (cqlType == null || cqlType.getTypeArguments().size() < 1)
+                        ? DataType.blob()
+                        : cqlType.getTypeArguments().get(0);
+                    DataType valueType = (cqlType == null || cqlType.getTypeArguments().size() < 2)
+                        ? DataType.blob() :
+                        cqlType.getTypeArguments().get(1);
+                    return newMapCodec(
+                        codecRegistry.codecFor(keyType, (TypeToken)null),
+                        codecRegistry.codecFor(valueType, (TypeToken)null));
+                } else {
+                    DataType keyType = (cqlType == null || cqlType.getTypeArguments().size() < 1)
+                        ? null
+                        : cqlType.getTypeArguments().get(0);
+                    DataType valueType = (cqlType == null || cqlType.getTypeArguments().size() < 2)
+                        ? null
+                        : cqlType.getTypeArguments().get(1);
+                    Map.Entry entry = (Map.Entry)map.entrySet().iterator().next();
+                    return (TypeCodec<T>)newMapCodec(
+                        codecRegistry.codecFor(keyType, entry.getKey()),
+                        codecRegistry.codecFor(valueType, entry.getValue()));
+                }
+            }
+
+            if ((cqlType == null || cqlType.getName() == DataType.Name.TUPLE) && value instanceof TupleValue) {
+                return (TypeCodec<T>)new TupleCodec(cqlType == null ? ((TupleValue)value).getType() : (TupleType)cqlType);
+            }
+
+            if ((cqlType == null || cqlType.getName() == DataType.Name.UDT) && value instanceof UDTValue) {
+                return (TypeCodec<T>)new UDTCodec(cqlType == null ? ((UDTValue)value).getType() : (UserType)cqlType);
+            }
+
+            if ((cqlType == null || cqlType.getName() == VARCHAR || cqlType.getName() == TEXT) && value instanceof Enum) {
+                return new EnumStringCodec(value.getClass());
+            }
+
+            if (cqlType != null && cqlType.getName() == INT && value instanceof Enum) {
+                return new EnumIntCodec(value.getClass());
+            }
+
+            return null;
+
+        }
+
+        /**
+         * Hook for subclasses wishing to customize codecs for CQL {@code list} types.
+         * @param eltCodec The element codec.
+         * @return A suitable codec.
+         */
+        protected <T> TypeCodec<List<T>> newListCodec(TypeCodec<T> eltCodec) {
```

Comment:
This would be the method to override if someone wants to create list codecs on the fly.